x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,302 @@
-	// if err != nil {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// }
-
-	// // If shares to migrate is not specified, we migrate all shares.
-	// if sharesToMigrate.IsZero() {
-	// 	sharesToMigrate = gammSharesInLock
-	// }
-
-	// // Otherwise, we must ensure that the shares to migrate is less than or equal to the shares in the lock.
-	// if sharesToMigrate.Amount.GT(gammSharesInLock.Amount) {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, fmt.Errorf("shares to migrate must be less than or equal to shares in lock")
-	// }
-
-	// // Exit the balancer pool position.
-	// exitCoins, err := k.gk.ExitPool(ctx, sender, poolIdLeaving, sharesToMigrate.Amount, sdk.NewCoins())
-	// if err != nil {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// }
-	// // Defense in depth, ensuring we are returning exactly two coins.
-	// if len(exitCoins) != 2 {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, fmt.Errorf("Balancer pool must have exactly two tokens")
-	// }
-
-	// // Create a full range (min to max tick) concentrated liquidity position.
-	// positionId, amount0, amount1, liquidity, joinTime, err = k.clk.CreateFullRangePosition(ctx, concentratedPool, sender, exitCoins)
-	// if err != nil {
-	// 	return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// }
-
-	// // If there are remaining gamm shares, we must re-lock them.
-	// remainingGammShares := gammSharesInLock.Sub(sharesToMigrate)
-	// if !remainingGammShares.IsZero() {
-	// 	newLock, err := k.lk.CreateLock(ctx, sender, sdk.NewCoins(remainingGammShares), remainingLockTime)
-	// 	newLockId = newLock.ID
-	// 	if err != nil {
-	// 		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// 	}
-	// 	// If the lock was unlocking, we begin the unlock from where it left off.
-	// 	if preUnlockLock.IsUnlocking() {
-	// 		_, err := k.lk.BeginForceUnlock(ctx, newLock.ID, newLock.Coins)
-	// 		if err != nil {
-	// 			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, err
-	// 		}
-	// 	}
-	// }
-
-	return positionId, amount0, amount1, liquidity, joinTime, poolIdLeaving, poolIdEntering, newLockId, nil
+func (k Keeper) MigrateLockedPositionFromBalancerToConcentrated(ctx sdk.Context, sender sdk.AccAddress, lockId uint64, sharesToMigrate sdk.Coin) (positionId uint64, amount0, amount1 sdk.Int, liquidity sdk.Dec, joinTime time.Time, poolIdLeaving, poolIdEntering, gammLockId, concentratedLockId uint64, err error) {
+	// Validate and retrieve pertinent data required for migration
+	poolIdLeaving, poolIdEntering, gammSharesInLock, concentratedPool, preMigrationLock, remainingLockTime, synthLockBeforeMigration, isSuperfluidBonded, isSuperfluidUnbonding, err := k.prepareMigration(ctx, sender, lockId, sharesToMigrate)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+
+	if isSuperfluidBonded {
```

Comment:
Would it be possible to have unit tests for these newly added three methods?
Please feel free to make an issue and handle it later instead if you feel this PR is getting to large

Reply:
I added an individual comment to each test case to show which sub method it runs. The top level function is simply a router. Please let me know if you still feel these each need their own test and I will make a separate issue:

https://github.com/osmosis-labs/osmosis/blob/70aee3f3a22424fe43f152605f5c9fe63c0fface/x/superfluid/keeper/migrate_test.go#L41-L72

Reply:
@czarcas7ic I would want to have each individual tdd test cases with separate cases handling all source of failure as they can come in handy if we change one of the lines in one of the methods in the future. if you agree please do make a separate issue, if you do not, please feel free to resolve this

Reply:
Sounds good, will add this as a low pri issue (can be done while audit is happening)

Reply:
Issue added here https://github.com/osmosis-labs/osmosis/issues/4977

Reply:
I don't think this is a low priority and should absolutely be tested pre-launch

Reply:
Low priority since it can be done during audit